srcCpp/Parameter.h

```diff
@@ -0,0 +1,289 @@
+#ifndef __PARAMETER_H__
+#define __PARAMETER_H__
+
+/*
+ * (c) 2005-2018  Copyright, Real-Time Innovations, Inc. All rights reserved.
+ * Subject to Eclipse Public License v1.0; see LICENSE.md for details.
+ */
+
+
+#include <stdio.h>
+#include <climits>
+#include <string>
+#include <utility>      // std::pair, std::make_pair
+#include <algorithm>    // std::sort
+#include <vector>
+#include <iostream>
+#include <sstream>
+
+enum TYPE {
+    T_NULL,           // Default type
+    T_NUMERIC,        // Numeric type, unsigned long long
+    T_STR,            // std::string tpy
+    T_BOOL,           // bool type
+    T_VECTOR_NUMERIC, // std::vector<unsigened long long>
+    T_VECTOR_STR,      // std::vector<std::string>
+    T_PAIR_NUMERIC_STR // std::pair<unsigened long long, std::string>
+};
+
+enum PARSEMETHOD {
+    NOSPLIT,          // If the value is not splited into the vector
+    SPLIT             // If the value is splited into the vector
+};
+
+// This specifies if a parameter will be followed by one extra argument.
+enum EXTRAARGUMENT {
+    NO,               // There is not extra argument
+    OPTIONAL,         // It is possible to have one extra argument
+    YES              // There is not one extra argument
+};
+
+enum GROUP {
+    GENERAL,
+    PUB,
+    SUB,
+    TRANSPORT,
+    SECURE,
+    RAWTRANSPORT
+};
+
+class ParameterBase  {
+    private:
+        std::pair <std::string, std::string> commandLineArgument;
+        std::string description;
+        bool isSet;
+        TYPE type;
+        EXTRAARGUMENT  extraArgument;
+        bool internal; // Does not have description
+        GROUP group;
+
+        // Only used for numeric argument
+        unsigned long long rangeStart;
+        unsigned long long rangeEnd;
+
+        // Only used for std::string argument
+        std::vector<std::string> validStrValues;
+
+    public:
+        ParameterBase();
+        virtual ~ParameterBase();
+
+        // Validate range
+        bool validate_numeric_range(unsigned long long var);
+
+        // Validate str Valuesi if not empty
+        bool validate_str_range(std::string var);
+
+        std::string print_command_line_parameter();
+
+        // Set members
+        virtual void set_command_line_argument(std::pair <std::string, std::string> var);
```

Comment:
I don’t think that the `virtual` statement is needed for those functions that don't gonna be overloaded by a child `class`